hls-plugin-api/src/Ide/Plugin/Properties.hs

```diff
@@ -0,0 +1,341 @@
+  ( PropertyType (..),
+    ToHsType,
+    MetaData (..),
+    PropertyKey,
+    SPropertyKey (..),
+    Properties,
+    HasProperty,
+    emptyProperties,
+    defineNumberProperty,
+    defineStringProperty,
+    defineBooleanProperty,
+    defineObjectProperty,
+    defineArrayProperty,
+    defineEnumProperty,
+    toDefaultJSON,
+    toVSCodeExtensionSchema,
+    usePropertyEither,
+    useProperty,
+    (&),
+  )
+where
+
+import qualified Data.Aeson       as A
+import qualified Data.Aeson.Types as A
+import           Data.Either      (fromRight)
+import           Data.Function    ((&))
+import           Data.Kind        (Constraint)
+import qualified Data.Map.Strict  as Map
+import           Data.Proxy       (Proxy (..))
+import qualified Data.Text        as T
+import           GHC.TypeLits
+import           Unsafe.Coerce    (unsafeCoerce)
+
+-- ---------------------------------------------------------------------
+
+data PropertyType
+  = TNumber
+  | TString
+  | TBoolean
+  | TObject
+  | TArray
+  | TEnum
+
+type family ToHsType (t :: PropertyType) where
+  ToHsType 'TNumber = Int
+  ToHsType 'TString = T.Text
+  ToHsType 'TBoolean = Bool
+  ToHsType 'TObject = A.Object
+  ToHsType 'TArray = A.Array
+  ToHsType 'TEnum = T.Text -- supports only text enum now
+
+-- ---------------------------------------------------------------------
+
+data MetaData (t :: PropertyType) where
+  MetaData ::
+    (IsNotTEnum t) =>
+    {defaultValue :: ToHsType t, description :: T.Text} ->
+    MetaData t
+  EnumMetaData ::
+    (IsTEnum t) =>
+    { defaultValue :: ToHsType t,
+      description :: T.Text,
+      enumValues :: [T.Text],
+      enumDescriptions :: [T.Text]
+    } ->
+    MetaData t
+
+data PropertyKey (s :: Symbol) (t :: PropertyType)
+
+data SPropertyKey k where
+  SNumber :: SPropertyKey (PropertyKey s 'TNumber)
+  SString :: SPropertyKey (PropertyKey s 'TString)
+  SBoolean :: SPropertyKey (PropertyKey s 'TBoolean)
+  SObject :: SPropertyKey (PropertyKey s 'TObject)
+  SArray :: SPropertyKey (PropertyKey s 'TArray)
+  SEnum :: SPropertyKey (PropertyKey s 'TEnum)
+
+data SomePropertyKeyWithMetaData
+  = forall k s t.
+    (k ~ PropertyKey s t) =>
+    SomePropertyKeyWithMetaData (SPropertyKey k, MetaData t)
+
+-- | Describes dedicated configuration of a plugin.
+--
+-- It was designed to be compatible with vscode's poor settings UI
+newtype Properties (r :: [*]) = Properties (Map.Map String SomePropertyKeyWithMetaData)
+
+-- ---------------------------------------------------------------------
+
+type family IsTEnum (t :: PropertyType) :: Constraint where
+  IsTEnum 'TEnum = ()
+  IsTEnum x = TypeError ('Text "Expected ‘" ':<>: 'ShowType 'TEnum ':<>: 'Text "’, but got ‘" ':<>: 'ShowType x ':<>: 'Text "’")
+
+type family IsNotTEnum (t :: PropertyType) :: Constraint where
+  IsNotTEnum 'TEnum = TypeError ('Text "Unexpected " ':<>: 'ShowType 'TEnum)
+  IsNotTEnum x = ()
+
+type family FindKey (s :: Symbol) r where
```

Comment:
Please give a kind on `r` here and in all of these type families.

Reply:
`r` is a list of `PropertyKey`, so I guess it has kind `[*]`?